internal/controller/template_controller.go

```diff
@@ -119,98 +171,90 @@ func (r *TemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
+			status.Providers.InfrastructureProviders = strings.Split(infraProviders, ",")
 		}
 	}
-	if len(template.Spec.Providers.BootstrapProviders) > 0 {
-		template.Status.Providers.BootstrapProviders = template.Spec.Providers.BootstrapProviders
+	if len(spec.Providers.BootstrapProviders) > 0 {
+		status.Providers.BootstrapProviders = spec.Providers.BootstrapProviders
 	} else {
 		bootstrapProviders := chart.Metadata.Annotations[hmc.ChartAnnotationBootstrapProviders]
 		if bootstrapProviders != "" {
-			template.Status.Providers.BootstrapProviders = strings.Split(bootstrapProviders, ",")
+			status.Providers.BootstrapProviders = strings.Split(bootstrapProviders, ",")
 		}
 	}
-	if len(template.Spec.Providers.ControlPlaneProviders) > 0 {
-		template.Status.Providers.ControlPlaneProviders = template.Spec.Providers.ControlPlaneProviders
+	if len(spec.Providers.ControlPlaneProviders) > 0 {
+		status.Providers.ControlPlaneProviders = spec.Providers.ControlPlaneProviders
 	} else {
 		cpProviders := chart.Metadata.Annotations[hmc.ChartAnnotationControlPlaneProviders]
 		if cpProviders != "" {
-			template.Status.Providers.ControlPlaneProviders = strings.Split(cpProviders, ",")
+			status.Providers.ControlPlaneProviders = strings.Split(cpProviders, ",")
 		}
 	}
 	return nil
 }
 
-func (r *TemplateReconciler) updateStatus(ctx context.Context, template *hmc.Template, validationError string) error {
-	template.Status.ObservedGeneration = template.Generation
-	template.Status.ValidationError = validationError
-	template.Status.Valid = validationError == ""
-	if err := r.Status().Update(ctx, template); err != nil {
-		return fmt.Errorf("failed to update status for template %s/%s: %w", template.Namespace, template.Name, err)
+func (r *TemplateReconciler) updateStatus(ctx context.Context, template hmc.Template, templateState *hmc.TemplateState, validationError string) error {
+	templateState.Status.ObservedGeneration = templateState.Generation
+	templateState.Status.ValidationError = validationError
+	templateState.Status.Valid = validationError == ""
+	err := template.UpdateStatus(ctx, r.Client, templateState.Status)
+	if err != nil {
+		return fmt.Errorf("failed to update status for template %s/%s: %w", templateState.Namespace, templateState.Name, err)
 	}
 	return nil
 }
 
-func (r *TemplateReconciler) reconcileHelmChart(ctx context.Context, template *hmc.Template) (*sourcev1.HelmChart, error) {
-	if template.Spec.Helm.ChartRef != nil {
+func (r *TemplateReconciler) reconcileHelmChart(ctx context.Context, templateState hmc.TemplateState) (*sourcev1.HelmChart, error) {
+	if templateState.Spec.Helm.ChartRef != nil {
 		// HelmChart is not managed by the controller
 		return nil, nil
 	}
 	helmChart := &sourcev1.HelmChart{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      template.Name,
-			Namespace: template.Namespace,
+			Name:      templateState.ObjectMeta.Name,
+			Namespace: r.SystemNamespace,
```

Comment:
I think we want to use the namespace of associated Cluster/ServiceTemplate object here. SystemNamespace is only for ProviderTemplate (when Namespace is empty)